server/sonar-web/src/main/js/apps/webhooks/components/CreateWebhookForm.tsx

```diff
@@ -0,0 +1,193 @@
+import Modal from '../../../components/controls/Modal';
+import ModalValidationField from '../../../components/controls/ModalValidationField';
+import { Webhook } from '../../../app/types';
+import { translate } from '../../../helpers/l10n';
+
+interface Props {
+  onClose: () => void;
+  onDone: (data: { name: string; url: string }) => Promise<void>;
+  webhook?: Webhook;
+}
+
+interface State {
+  name: string;
+  nameError?: string;
+  submitting: boolean;
+  url: string;
+  urlError?: string;
+}
+
+export default class CreateWebhookForm extends React.PureComponent<Props, State> {
+  mounted: boolean;
+
+  constructor(props: Props) {
+    super(props);
+    const { webhook } = props;
+    const name = webhook ? webhook.name : '';
+    const url = webhook ? webhook.url : '';
+    this.state = {
+      name,
+      nameError: this.validateName(name),
+      submitting: false,
+      url,
+      urlError: this.validateUrl(url)
+    };
+  }
+
+  componentDidMount() {
+    this.mounted = true;
+  }
+
+  componentWillUnmount() {
+    this.mounted = false;
+  }
+
+  handleCancelClick = (event: React.SyntheticEvent<HTMLButtonElement>) => {
+    event.preventDefault();
+    event.currentTarget.blur();
+    this.props.onClose();
+  };
+
+  handleNameChange = (name: string) => this.setState({ name, nameError: this.validateName(name) });
+
+  handleUrlChange = (url: string) => this.setState({ url, urlError: this.validateUrl(url) });
+
+  handleFormSubmit = (event: React.SyntheticEvent<HTMLFormElement>) => {
+    event.preventDefault();
+    if (this.isValid()) {
+      this.setState({ submitting: true });
+      this.props
+        .onDone({ name: this.state.name, url: this.state.url })
+        .then(this.props.onClose, () => {
+          if (this.mounted) {
+            this.setState({ submitting: false });
+          }
+        });
+    }
+  };
+
+  isValid = () => !this.state.nameError && !this.state.urlError;
+
+  validateName = (name: string) => {
+    if (!name.trim()) {
+      return translate('webhooks.name.required');
+    }
+    return undefined;
+  };
+
+  validateUrl = (url: string) => {
+    if (!url.trim()) {
+      return translate('webhooks.url.required');
+    }
+    if (!url.startsWith('http://') && !url.startsWith('https://')) {
+      return translate('webhooks.url.bad_protocol');
+    }
+    if (url.indexOf(':', 6) > 0 && url.indexOf('@') <= 0) {
+      return translate('webhooks.url.bad_auth');
+    }
+  };
+
+  render() {
+    const { submitting } = this.state;
+    const isUpdate = !!this.props.webhook;
+    const modalHeader = isUpdate ? translate('webhooks.update') : translate('webhooks.create');
+
+    return (
+      <Modal contentLabel={modalHeader} onRequestClose={this.props.onClose}>
+        <form onSubmit={this.handleFormSubmit}>
+          <div className="modal-head">
+            <h2>{modalHeader}</h2>
+          </div>
+
+          <div className="modal-body">
+            <ModalValidationField
```

Comment:
Could you extract this to something like: 

```typescript
<TextInputField 
  id={...} 
  label={...} 
  onChange={this.handleInputChange} 
  onError={this.disableForm}
  required={true} />
``` 

I'd really love to encapsulate the validation process inside the component, and to only know if the field is valid or not.